Can not compile when using `WITH_STL=CXX11` with legacy toolchain

**Describe your environment** 

Sorry but #3098 only fixes partly problems in #3096 . I will raise another PR to fix the left problems.

OS: Linux
Compiler: GCC 4.8.5
otel-cpp version: 1.17.0

**Steps to reproduce**
Building otel-cpp 1.17.0

**What is the expected behavior?**
Building success.

**What is the actual behavior?**

```bash
// ...
/home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.17.0/exporters/memory/src/in_memory_metric_exporter_factory.cc:82:26: error: parameter declared ‘auto’
   return Create(data, [](auto) { return AggregationTemporality::kCumulative; });


// ...
/home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.17.0/exporters/memory/src/in_memory_metric_data.cc: In member function ‘virtual void opentelemetry::v1::exporter::memory::SimpleAggregateInMemoryMetricData::Add(std::unique_ptr<opentelemetry::v1::sdk::metrics::ResourceMetrics>)’:
/home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.17.0/exporters/memory/src/in_memory_metric_data.cc:34:30: error: converting to ‘std::map<std::tuple<std::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::map<opentelemetry::v1::sdk::common::OrderedAttributeMap, absl::lts_2020_02_25::variant<opentelemetry::v1::sdk::metrics::SumPointData, opentelemetry::v1::sdk::metrics::HistogramPointData, opentelemetry::v1::sdk::metrics::LastValuePointData, opentelemetry::v1::sdk::metrics::DropPointData> > >::key_type {aka std::tuple<std::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::basic_string<char, std::char_traits<char>, std::allocator<char> > >}’ from initializer list would use explicit constructor ‘constexpr std::tuple<_T1, _T2>::tuple(const _T1&, const _T2&) [with _T1 = std::basic_string<char>; _T2 = std::basic_string<char>]’
         data_[{scope, metric}].insert({pda.attributes, pda.point_data});
                              ^
/home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.17.0/exporters/memory/src/in_memory_metric_data.cc: In member function ‘const AttributeToPoint& opentelemetry::v1::exporter::memory::SimpleAggregateInMemoryMetricData::Get(const string&, const string&)’:
/home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.17.0/exporters/memory/src/in_memory_metric_data.cc:44:31: error: converting to ‘std::map<std::tuple<std::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::map<opentelemetry::v1::sdk::common::OrderedAttributeMap, absl::lts_2020_02_25::variant<opentelemetry::v1::sdk::metrics::SumPointData, opentelemetry::v1::sdk::metrics::HistogramPointData, opentelemetry::v1::sdk::metrics::LastValuePointData, opentelemetry::v1::sdk::metrics::DropPointData> > >::key_type {aka std::tuple<std::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::basic_string<char, std::char_traits<char>, std::allocator<char> > >}’ from initializer list would use explicit constructor ‘constexpr std::tuple<_T1, _T2>::tuple(const _T1&, const _T2&) [with _T1 = std::basic_string<char>; _T2 = std::basic_string<char>]’
   return data_[{scope, metric}];                    ^
```

--- a/exporters/memory/src/in_memory_metric_data.cc
+++ b/exporters/memory/src/in_memory_metric_data.cc
@@ -31,7 +31,8 @@ void SimpleAggregateInMemoryMetricData::Add(std::unique_ptr<ResourceMetrics> res
       const auto &metric = m.instrument_descriptor.name_;
       for (const auto &pda : m.point_data_attr_)
       {
-        data_[{scope, metric}].insert({pda.attributes, pda.point_data});
+        data_[std::tuple<std::string, std::string>{scope, metric}].insert(
+            {pda.attributes, pda.point_data});
       }
     }
   }
@@ -41,7 +42,7 @@ const SimpleAggregateInMemoryMetricData::AttributeToPoint &SimpleAggregateInMemo
     const std::string &scope,
     const std::string &metric)
 {
-  return data_[{scope, metric}];
+  return data_[std::tuple<std::string, std::string>{scope, metric}];
 }
 
 void SimpleAggregateInMemoryMetricData::Clear()

--- a/exporters/memory/src/in_memory_metric_data.cc
+++ b/exporters/memory/src/in_memory_metric_data.cc
@@ -31,6 +31,7 @@ void SimpleAggregateInMemoryMetricData::Add(std::unique_ptr<ResourceMetrics> res
       const auto &metric = m.instrument_descriptor.name_;
       for (const auto &pda : m.point_data_attr_)
       {
+        // NOTE: Explicit type conversion added for C++11 (gcc 4.8)
         data_[std::tuple<std::string, std::string>{scope, metric}].insert(
             {pda.attributes, pda.point_data});
       }
@@ -42,6 +43,7 @@ const SimpleAggregateInMemoryMetricData::AttributeToPoint &SimpleAggregateInMemo
     const std::string &scope,
     const std::string &metric)
 {
+  // NOTE: Explicit type conversion added for C++11 (gcc 4.8)
   return data_[std::tuple<std::string, std::string>{scope, metric}];
 }
 

--- a/exporters/memory/src/in_memory_metric_exporter_factory.cc
+++ b/exporters/memory/src/in_memory_metric_exporter_factory.cc
@@ -78,7 +78,8 @@ private:
 std::unique_ptr<PushMetricExporter> InMemoryMetricExporterFactory::Create(
     const std::shared_ptr<InMemoryMetricData> &data)
 {
-  return Create(data, [](auto) { return AggregationTemporality::kCumulative; });
+  return Create(data,
+                [](sdk::metrics::InstrumentType) { return AggregationTemporality::kCumulative; });
 }
 
 std::unique_ptr<PushMetricExporter> InMemoryMetricExporterFactory::Create(

